File: go/day-6/p1.go
package main

import (
	"Advent-of-Code-2023/utils"
	"fmt"
	"strings"
)

type Pair struct {
	T, D int // time distance
}

func PartOne() {
	ans := 1
	lines := utils.ReadLines("input.txt")
	pairs := ToPairs(lines)

	for _, p := range pairs {
		c := calc(p.T, p.D)
		ans *= c
	}
	fmt.Println("ans: ", ans)
}

func calc(tToBeat, dToBeat int) int {
	var c int
	for hold := 1; hold < tToBeat; hold++ {
		dRes := (tToBeat - hold) * hold
		if dRes > dToBeat {
			c++
		}
	}
	return c
}

func ToPairs(lines []string) []Pair {
	var pairs []Pair
	// parse
	times, dists := ParseLine(lines[0]), ParseLine(lines[1])
	size := len(times)
	for i := 0; i < size; i++ {
		pair := Pair{
			T: times[i],
			D: dists[i],
		}
		pairs = append(pairs, pair)
	}
	return pairs
}

func ParseLine(line string) []int {
	var l []int
	nums := strings.Fields(line)[1:]
	for _, n := range nums {
		l = append(l, utils.StrToInt(n))
	}
	return l
}

File: python/day-8/p1.py
def do():
    init, rules = parse()
    lost, at, c = True, "AAA", 0

    while lost:
        for d in init:
            c += 1
            if d == "L":
                at = rules[at][0]
            elif d == "R":
                at = rules[at][1]

            if at == "ZZZ":
                lost = False

    print(f"ans: {c}")


def parse() -> (list, dict):
    with open("input.txt") as f:
        lines = f.readlines()

    init = lines[0].strip()
    rules = {}

    for line in lines[2:]:
        split = line.strip().split(" = ")
        col2 = split[1][1:-1].split(", ")
        rules[split[0]] = col2
    return init, rules


if __name__ == "__main__":
    do()
